### ISSUE TITLE
gss_accept_sec_context() failed: Unknown error

### ISSUE THREAD

**Issue by ermacaz** [stnoonan/spnego-http-auth-nginx-module#60]
Hi,

I am getting this error in the nginx logs when attempting to authenticate. Below is as much relevant logs/settings I could think of.  Any help is appreciated.

nginx log
````
2016/04/01 19:35:08 [debug] 22415#0: *4 http header: "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
2016/04/01 19:35:08 [debug] 22415#0: *4 http header: "Accept-Language: en-US,en;q=0.5"
2016/04/01 19:35:08 [debug] 22415#0: *4 http header: "Accept-Encoding: gzip, deflate"
2016/04/01 19:35:08 [debug] 22415#0: *4 http header: "Connection: keep-alive"
2016/04/01 19:35:08 [debug] 22415#0: *4 http header: "Cache-Control: max-age=0"
2016/04/01 19:35:08 [debug] 22415#0: *4 http header: "Authorization: Negotiate TlRMTVNTUAABAAAAl4II4gAAAAAAAAAAAAAAAAAAAAAGAbEdAAAADw=="
2016/04/01 19:35:08 [debug] 22415#0: *4 http header done
2016/04/01 19:35:08 [debug] 22415#0: *4 rewrite phase: 0
2016/04/01 19:35:08 [debug] 22415#0: *4 test location: "/"
2016/04/01 19:35:08 [debug] 22415#0: *4 using configuration "/"
2016/04/01 19:35:08 [debug] 22415#0: *4 http cl:-1 max:52428800
2016/04/01 19:35:08 [debug] 22415#0: *4 rewrite phase: 2
2016/04/01 19:35:08 [debug] 22415#0: *4 post rewrite phase: 3
2016/04/01 19:35:08 [debug] 22415#0: *4 generic phase: 4
2016/04/01 19:35:08 [debug] 22415#0: *4 generic phase: 5
2016/04/01 19:35:08 [debug] 22415#0: *4 access phase: 6
2016/04/01 19:35:08 [debug] 22415#0: *4 SSO auth handling IN: token.len=0, head=0, ret=401
2016/04/01 19:35:08 [debug] 22415#0: *4 Begin auth
2016/04/01 19:35:08 [debug] 22415#0: *4 Detect SPNEGO token
2016/04/01 19:35:08 [debug] 22415#0: *4 posix_memalign: 000000000198B100:4096 @16
2016/04/01 19:35:08 [debug] 22415#0: *4 Token decoded: TlRMTVNTUAABAAAAl4II4gAAAAAAAAAAAAAAAAAAAAAGAbEdAAAADw==
2016/04/01 19:35:08 [debug] 22415#0: *4 Client sent a reasonable Negotiate header
2016/04/01 19:35:08 [debug] 22415#0: *4 GSSAPI authorizing
2016/04/01 19:35:08 [debug] 22415#0: *4 Use keytab /etc/krb5.keytab
2016/04/01 19:35:08 [debug] 22415#0: *4 gss_accept_sec_context() failed: Unknown error:
2016/04/01 19:35:08 [debug] 22415#0: *4 GSSAPI failed
2016/04/01 19:35:08 [debug] 22415#0: *4 http finalize request: 403, "/?" a:1, c:1
2016/04/01 19:35:08 [debug] 22415#0: *4 http special response: 403, "/?"
2016/04/01 19:35:08 [debug] 22415#0: *4 http set discard body
2016/04/01 19:35:08 [debug] 22415#0: *4 HTTP/1.1 403 Forbidden
Server: nginx/1.4.4
````
Excerpt from nginx conf
````
location / {
    add_header REMOTE_USER $remote_user;
    auth_gss on;
    auth_gss_keytab /etc/krb5.keytab;
 }
````

klist -e output
```
Ticket cache: FILE:/tmp/krb5cc_1005
Default principal: HTTP/ad.hidden.com@HIDDEN.COM

Valid starting       Expires              Service principal
04/01/2016 19:44:48  04/02/2016 05:44:48  krbtgt/HIDDEN.COM@HIDDEN.COM
	renew until 04/02/2016 19:44:48, Etype (skey, tkt): arcfour-hmac, arcfour-hmac
```

klist -kte output
```
Keytab name: FILE:/etc/krb5.keytab
KVNO Timestamp           Principal
---- ------------------- ------------------------------------------------------
   2 04/01/2016 18:44:57 HTTP/ad.hidden.com@HIDDEN.COM (arcfour-hmac)
   3 04/01/2016 19:43:25 HTTP/ad.hidden.com@HIDDEN.COM (arcfour-hmac)
```
permissions on krb5.keytab/.conf (nginx runs as www-data user)
```
-rw-rw----  1 www-data root     953 Apr  1 19:31 krb5.conf
-rw-rw----  1 www-data root     158 Apr  1 19:43 krb5.keytab
```

krb5.conf
```
[libdefaults]
    #default_tkt_enctypes = arcfour-hmac-md5
    #default_tgs_enctypes = arcfour-hmac-md5
    #default_keytab_name  = FILE:/etc/krb5.keytab
    default_tgs_enctypes = rc4-hmac arcfour-hmac-md5 des-cbc-crc des-cbc-md5
    default_tkt_enctypes = rc4-hmac arcfour-hmac-md5 des-cbc-crc des-cbc-md5
    permitted_enctypes = rc4-hmac arcfour-hmac-md5 arcfour-hmac
    default_realm        = HIDDEN.COM
    ticket_lifetime      = 24h
    kdc_timesync         = 1
    ccache_type          = 4
    forwardable          = false
    proxiable            = false

  [realms]
    HIDDEN.COM = {
        database_name = /var/lib/krb5kdc/principal
        kdc            = kdc.hidden.com
        admin_server   = kdc.hidden.com
        default_domain = hidden.com
    }

  [domain_realm]
    .kerberos.server = hidden.COM
    .hidden.com     = hidden.COM
    hidden.com = hidden.COM

[logging]
       kdc = FILE:/var/log/krb5kdc.log
```
kdc.conf
```
[kdcdefaults]
    kdc_ports = 750,88

[realms]
    HIDDEN.COM = {
        database_name = /var/lib/krb5kdc/principal
        admin_keytab = FILE:/etc/krb5.keytab
        acl_file = /etc/krb5kdc/kadm5.acl
        key_stash_file = /etc/krb5kdc/stash
        kdc_ports = 750,88
        max_life = 10h 0m 0s
        max_renewable_life = 7d 0h 0m 0s
        master_key_type = rc4-hmac
        supported_enctypes = rc4-hmac:normal arcfour-hmac:normal arcfour-hmac-md5:normal
        default_principal_flags = +preauth
    }

[logging]
                kdc = FILE:/tmp/kdc.log
                admin_server = FILE:/tmp/kadmin.log
```

**Comment by stnoonan** [comment]
Client is misconfigured.  It is trying to NTLM.

echo -n "TlRMTVNTUAABAAAAl4II4gAAAAAAAAAAAAAAAAAAAAAGAbEdAAAADw==" | base64
-d

NTLMSSP �� �

**Comment by ermacaz** [comment]
Thank you, that pointed me in the right direction.  Was able to fix my issue after regenerating the HTTP keytab entry.

**Comment by xayyuu** [comment]
@ermacaz 

 I met the same error( whether using firefox or Internet Explorer).
thank you for your  issue, i know that the error maybe 'Client is misconfigured. It is trying to NTLM.'
i can not find how to avoid trying to NTLM, how do you slove this problem at last ?

coud you give me some suggest?

**Comment by offtza01** [comment]
*bump*
Same here - what do you mean - "Client is misconfigured. It is trying to NTLM."
Is there any way to force basic token configuration?

**Comment by TiTi87** [comment]
@ermacaz Could you detail how you were able to fix this with "regenerating the HTTP keytab entry" ?

I know my server config is fine since "auth_gss_allow_basic_fallback" is working perfectly. But SSO is failing with "gss_accept_sec_context() failed: Unknown error" because the client is sending an NTLM ticket.

Thank in advance.

**Comment by ermacaz** [comment]
@TiTi87 These are from my notes from when I got it working back then.
1. Created Active Directory user with following config
```
User cannot change password
Password never expires
Does not require Kerberos pre-authentication
```
2.  Create keytab files with the following commands in cmd on machine with Active Directory.  Replace fields in [] with real values and remove [].
```
ktpass -princ host/[URL]@[DOMAIN] -mapuser   AD\[AD_USERNAME] -pass '[AD_USER_PASSWORD]' -out host.keytab -ptype KRB5_NT_PRINCIPAL

ktpass -princ HTTP/[URL]@[DOMAIN] -mapuser   AD\[AD_USERNAME] -pass '[AD_USER_PASSWORD]' -out http.keytab -ptype KRB5_NT_PRINCIPAL
```
3.  Copy keytab files to webserver and use ktutil to combine and save as /etc/krb5.keytab.  Ensure permissions are set so it is readable by  the webserver 
```
sudo ktutil 
ktutil:  rkt host.keytab 
ktutil:  rkt http.keytab 
ktutil:  wkt /etc/krb5.keytab 
ktutil:  quit
sudo chown WEB_SERVER_PROCESS_USER /etc/krb5.keytab
sudo chmod 600 /etc/krb5.keytab
```
4.  Restart webserver.
5.  Ensure client's browser is configured properly, details here
https://www.ibm.com/support/knowledgecenter/en/SSEQTP_9.0.5/com.ibm.websphere.base.doc/ae/tsec_SPNEGO_configweb_new.html

**Comment by TiTi87** [comment]
Thank @ermacaz, 
Your inputs allowed me to confirm that my webserver and clients configurations are correct.
I suspect that the problem is somewhere else in our AD security/GPO settings blocking the client in NTLM mode. I will try to post here if i found a way around this...
Thank again.

**Comment by TiTi87** [comment]
Ok so, I finally sorted it out.
It turned out that my configuration was fine... but that the DNS configuration of the server was unusual. The URL I was given for the webserver was in fact a DNS alias (CNAME), not the "A record" for this machine.
Kinit and spnego-module seems to handle it just fine as we specify witch servicename to use... but the clients (firefox, ie, chrome) does not ([src 1](https://blogs.technet.microsoft.com/askds/2008/05/29/kerberos-authentication-problems-service-principal-name-spn-issues-part-1/), [src 2](https://support.pingidentity.com/s/article/How-browsers-construct-Kerberos-requests-service-principal-names)), a few network captures later, shows that Firefox was asking a TGS with the "A-Record DNS server name" (i.e HTTP/A-record-FQDN@DOMAIN.TLD), to the KDC, not the alias.
So first problem the KDC can't return any TGS, and the client falls back to NTLM Auth.
Second issue, even if the KDC returned a valid TGS, my service is using the alias name for the Keytab, so the TGS would'nt have been valid for this service.
Long story short, a few files permission changes on the keetab and the large_client_header_buffers param set...
Hope it will help someone.
Thank again @ermacaz, and everyone in this thread for your help.

**Comment by alokkumar0009** [comment]
I am also facing similar issue: [gss_accept_sec_context() failed: Unknown error, Is it required to generate two keytab file, for host and http? Can I not run my nginx proxy using root, in that case what permission is requied for keytab file?
